src/repr/src/adt/interval.rs

```diff
@@ -272,16 +294,57 @@ impl Interval {
         (self.micros % 1_000_000 * 1_000).try_into().unwrap()
     }
 
-    /// Computes the total number of seconds in the interval.
-    pub fn as_seconds<T>(&self) -> T
+    /// Computes the total number of epoch seconds in the interval.
+    /// When extracting an epoch, PostgreSQL considers a year
+    /// 365.25 days.
+    pub fn as_epoch_seconds<T>(&self) -> T
     where
         T: DecimalLike,
     {
-        T::from(self.years() * 60 * 60 * 24) * T::from(365.25)
-            + T::from(self.months() * 60 * 60 * 24 * 30)
-            + T::from(self.days * 60 * 60 * 24)
-            + T::lossy_from(self.dur_as_secs())
-            + T::from(self.nanoseconds()) / T::from(1e9)
+        let days = T::from(self.years()) * T::from(Self::EPOCH_DAYS_PER_YEAR)
+            + T::from(self.months()) * T::from(Self::DAY_PER_MONTH)
+            + T::from(self.days);
+        let seconds =
+            days * T::from(Self::HOUR_PER_DAY * Self::MINUTE_PER_HOUR * Self::SECOND_PER_MINUTE);
+
+        seconds
+            + T::lossy_from(self.micros)
+                / T::from(Self::MICROSECOND_PER_MILLISECOND * Self::MILLISECOND_PER_SECOND)
+    }
+
+    /// Computes the total number of microseconds in the interval.
+    pub fn as_microseconds(&self) -> i128 {
+        // unwrap is safe because i32::MAX/i32::MIN number of months will not overflow an i128 when
+        // converted to microseconds.
+        Self::convert_date_time_unit(
+            DateTimeField::Month,
+            DateTimeField::Microseconds,
+            i128::from(self.months),
+        ).unwrap() +
+        // unwrap is safe because i32::MAX/i32::MIN number of days will not overflow an i128 when
+        // converted to microseconds.
+        Self::convert_date_time_unit(
+            DateTimeField::Day,
+            DateTimeField::Microseconds,
+            i128::from(self.days),
+        ).unwrap() +
+        i128::from(self.micros)
+    }
+
+    /// Converts this `Interval`'s duration into `chrono::Duration`.
+    pub fn duration_as_chrono(&self) -> chrono::Duration {
+        use chrono::Duration;
+        Duration::days(self.days.into()) + Duration::microseconds(self.micros)
```

Comment:
Still scared this doesn't have months in it. Should assert or something.

Reply:
I'm just keeping this consistent with the previous implementation which used both days and microseconds, but didn't use months. Honestly not sure right now if it was a bug that the previous implementation didn't use months though.

@sploiselle any thoughts? This is used in the following functions

- `add_date_interval`
- `add_time_interval`
- `add_timestamplike_interval`
- `sub_date_interval`
- `sub_time_interval`
- `timezone_interval_time`
- `timezone_interval_timestamp`
- `timezone_interval_timestamptz`

If this is an existing bug, I'm somewhat inclined to maintain the previous behavior and fix the bug in a followup issue/PR since this PR has grown pretty big. Thought?

Reply:
Ah, sorry I missed this earlier; this simply doesn't have months in it because month addition was handled by a separate function, e.g. `add_timestamp_months`. I think we can tidy up how all of this works in a separate, follow-on PR. Essentially, we just need to figure out what the semantics are of adding unit-crossing values (e.g. 100 days, 100 months) to these other units. As long as this implementation holds the same semantics as the previous implementation, I think it's fine to move to follow-on work as long as there's bandwidth to do it immediately.

If you plan to go back through CI, @jkosh44, add a `TODO` here.

Reply:
Just added a TODO to the function